### ISSUE TITLE
Init process in SMP PAR mode is very slow

### ISSUE THREAD

**Issue by kilaterlee** [memtest86plus/memtest86plus#403]
Hi,
I have a problem that I found it is very slow in init process when in SMP(PAR) mode with 256G memory 64 cores, I checked the code, it seems each core will runs the sample test twice, and each sample case iterates through all of the address windows. 
So if it has 64 core-256G memory, I test it all night and still didn't pass the init process.
Has anyone encountered the same problem?

**Comment by debrouxl** [comment]
My computers with 64 cores and 256 GB of RAM normally spend several seconds in the init phase. All night without proceeding to test #0 shouldn't happen, though.
First of all: which version of memtest86+ are you using: 7.00 release, nightly version, something else ?

**Comment by kilaterlee** [comment]
> My computers with 64 cores and 256 GB of RAM normally spend several seconds in the init phase. All night without proceeding to test #0 shouldn't happen, though. First of all: which version of memtest86+ are you using: 7.00 release, nightly version, something else ? Also, could you give some details about your platform: for instance, processor model, number of sockets, possibly commercial computer model name ?

Thanks for you reply, I'm using the code after V7 and the comiit id is ded371e9dafcc.
However, I'm using a non-x64 platform to test with SMP PAR mode. Currently, I know that the memtest86+ does not allow other patform testing, but I'm one of LoongArch maintainers in edk2 and edk2-platfroms, I'm also work at Loongson, so it is easier for me to add the LoongArch suppot into memtest86+, and I promise the code added just for LoongArch supports it will not affect other test logic.
This a other topic that whether memtest86+ going to accommodate more ARCH or whether add multi-ARCH support. I hope it can become possible after my patch, I think I took into account the rules the multi-ARCH in memtest86+, and the architecture related code should be independent from the test logic code.
That should be:
1. All of the API or UI related should be public, and everyone must follow the public API, such as cpu info, clock, IMC info, flush cache, AMO function, etc.
2. ARCH init and the ARCH related codes should be private and independent.
3. The test logic is public and does not depend on any feature or ARCH. It is looking good now.

After all, the memtest86+ can made more powerful and can accommodate more ARCH. If I submit the reivew PRs, would you be interested in reviewing them?

**Comment by x86fr** [comment]
Yes! Also, we'll be very interested to add Longson Arch on all the detection/ecc polling code. Are datasheets public?

**Comment by martinwhitaker** [comment]
Does the boot option `powersave=off` make any difference?

With trace enabled (either through the F1 config menu or using the boot option `trace`), what is displayed?

You can add more detailed tracing by setting `TRACE_BARRIERS` to `1` in main.c.

**Comment by kilaterlee** [comment]
> Yes! Also, we'll be very interested to add Longson Arch on all the detection/ecc polling code. Are datasheets public?

Thank you! The code has not been cleaned up yet, I'm just porting done few days ago. If you guys are interested, I will make the repo public after code is cleaned up.

**Comment by kilaterlee** [comment]
> Does the boot option `powersave=off` make any difference?
I will try this option once the machine is free.
> 
> With trace enabled (either through the F1 config menu or using the boot option `trace`), what is displayed?
> 
> You can add more detailed tracing by setting `TRACE_BARRIERS` to `1` in main.c.
Yes, I enabled the trace, and I observed that the init process runs the dummy_run on different core one by one, and the windows will increase to 256G. 
So it looks like during the init phase, it is single threaded and iterates through all address windows in every dummy_run test.

**Comment by debrouxl** [comment]
Ah, a non-x86 arch.
Sure, please make the repo public when you have cleaned it up :)

If (certainly when) proper multi-arch support is added to memtest86+, then vendors of e.g. ARM- or RISC-V-based platforms can leverage it to contribute support for their (hopefully standardized enough, e.g. UEFI) platforms. On the maintainer side, time on the maintainer side and access to suitable platforms is way too limited for that.
#147 is the tracking issue of sorts for non-x86 ports, it links / is linked by other issues on this repo and other repos.

**Comment by kilaterlee** [comment]
> Ah, a non-x86 arch. Sure, please make the repo public when you have cleaned it up :)

Okay, I will do that as soon as possible.
> 
> If (certainly when) proper multi-arch support is added to memtest86+, then vendors of e.g. ARM- or RISC-V-based platforms can leverage it to contribute support for their (hopefully standardized enough, e.g. UEFI) platforms. On the maintainer side, time on the maintainer side and access to suitable platforms is way too limited for that. #147 is the tracking issue of sorts for non-x86 ports, it links / is linked by other issues on this repo and other repos.

Hmm, maybe like the edk2? Different ARCH related code maintained by the sub-maintainer?

**Comment by kilaterlee** [comment]
> Does the boot option `powersave=off` make any difference?
> 
> With trace enabled (either through the F1 config menu or using the boot option `trace`), what is displayed?
> 
> You can add more detailed tracing by setting `TRACE_BARRIERS` to `1` in main.c.

Hi martinwhitaker,
After I uses the `powersave=off`, it is faster than `powersave=high`, the 64 cores with 512G RAM takes about 7 minutes to init.

**Comment by martinwhitaker** [comment]
> After I uses the `powersave=off`, it is faster than `powersave=high`, the 64 cores with 512G RAM takes about 7 minutes to init.

That suggests that your machine doesn't support broadcasting a NMI to wake up halted threads (see the `barrier_halt_wait()` function in `lib/barrier.c`).

7 minutes is still a long time for the init phase. I would suggest looking into the performance of the `barrier_spin_wait()` function.

**Comment by kilaterlee** [comment]
> > After I uses the `powersave=off`, it is faster than `powersave=high`, the 64 cores with 512G RAM takes about 7 minutes to init.
> 
> That suggests that your machine doesn't support broadcasting a NMI to wake up halted threads (see the `barrier_halt_wait()` function in `lib/barrier.c`).
> 
> 7 minutes is still a long time for the init phase. I would suggest looking into the performance of the `barrier_spin_wait()` function.

I turned on the trace, and I observed that it runs the dummy_run on difference cores one by one, for example, core 0 runs the test0-10 then core 1 runs the test0-10 then core 2 ...
The barrier code I only added the ARCH related:
![image](https://github.com/memtest86plus/memtest86plus/assets/11434289/cddbbe2f-d3b1-4738-819a-8c70f712fa39)
![image](https://github.com/memtest86plus/memtest86plus/assets/11434289/e989c6d7-2d28-47aa-b65d-86053411a812)
BTW, the `barrier_halt_wait` looks like not uses the broadcasting, `spm_send_nmi` is in a loop and it looks like the last core will wakes up other cores one by one.

**Comment by kilaterlee** [comment]
Hi All,
Sorry for not updating for long time, I'm a little busy the past few days, and I'm still trying to find the root cause of the slowdown in the init phase of multi-core and large memory.
I think I've found some clues: the logic of finding PMEM in `test_all_windows()` is not friendly for the uniformly addressing ARCHs. In the uniformly addressing ARCHs, most PMEM are discontinuous. If the `pm_map[]` have some huge address hole between tow members, the init phase will be slow, for example: If the system contents two or more NUMA nodes, and each node has its own PMEM, usually they uses 4 or more bits to marks differents node address in uniformly addressing ARCHs(currently, [47:44] is used for mark the differents nodes in Loongson CPUs), so there maybe some huge address hole between two `pm_map[]` members. The logic to find PMEM is to use the `window_end` add the `VM_WINDOW_SIZE(1GB)` as the next window starts, in which case it will loop very many times to find the next valid address window while other cores just to wait-wake-wait-wake.
I'm still trying to find the perfect solution and minimal code modifications.

**Comment by debrouxl** [comment]
Interesting. Well, if the algorithm currently used by memtest86+ is the root cause of this very slow startup you're experiencing, then using another algorithm which handles your computer's situation better is acceptable.

Given a sufficiently advanced Loongson support in QEMU, the issue should be reproducible on emulated hardware, and debuggable through GDB, right ?
Even if the issue weren't reproducible with less than 256 GB of RAM allocated to the emulated hardware, a host computer with more than 256 GB of RAM isn't hard to find nowadays... how things have changed. This week, I stumbled across a discarded but working computer, equipped with 32 MB of RAM...

**Comment by kilaterlee** [comment]
> Interesting. Well, if the algorithm currently used by memtest86+ is the root cause of this very slow startup you're experiencing, then using another algorithm which handles your computer's situation better is acceptable.

I mapped the node number from bit 44 to bit 38, which is  much faster than before, I guess it is at least five times faster(or more). But this way will limits the max memory size of each NUMA node to 256G(BTW, I notice on my x86 servers it also use the bit 38 as the NUMA number, I guess it has the same issue if the node number was more higher).
> 
> Given a sufficiently advanced Loongson support in QEMU, the issue should be reproducible on emulated hardware, and debuggable through GDB, right ? Even if the issue weren't reproducible with less than 256 GB of RAM allocated to the emulated hardware, a host computer with more than 256 GB of RAM isn't hard to find nowadays... how things have changed. This week, I stumbled across a discarded but working computer, equipped with 32 MB of RAM...

Yes, I think it is should work fine on the QEMU, although I haven't to tried it yet, I will give a try next. But I don't GDB can help with this problem. :)

**Comment by kilaterlee** [comment]
@debrouxl I notice that you submitted a patch to initial the NUMA awareness on Mar 13. I guess that will help with this issue(My code is based the Mar 5 commit, so this change is not included in my code). I will learn this patch and adapt to this change when I porting and cleaning up my code. THX.

**Comment by martinwhitaker** [comment]
During the initial two dummy_run passes, memory is not being tested. So NUMA awareness should make no difference, because that memory is not being written or read.

**Comment by kilaterlee** [comment]
I also noticed that if I run multiple cores (maybe 256 cores, 512G RAM), and test 0 and test 2 are SEQ mode. In this mode, each core will iterate over all uncached addresses, and the test speed is pretty slow on the Loongson multi-node machine. Is there the same problem on X64?

**Comment by debrouxl** [comment]
Test 0 is usually fast, but tests 1 and 2 are indeed slow, and sensitive to NUMA effects. Across the range of 1-8 NUMA node x86 & x86-64 computers I tested them on, non-temporal writes (experimental `nontemporal` branch) usually speed up these tests, so they improve bandwidth and/or latency in those tests; however, they often slow down the other tests, even with loop unrolling.

**Comment by kilaterlee** [comment]
> Test 0 is usually fast, but tests 1 and 2 are indeed slow, and sensitive to NUMA effects. Across the range of 1-8 NUMA node x86 & x86-64 computers I tested them on, non-temporal writes (experimental `nontemporal` branch) usually speed up these tests, so they improve bandwidth and/or latency in those tests; however, they often slow down the other tests, even with loop unrolling.

It seems the test 1 is falsed by default:
![image](https://github.com/memtest86plus/memtest86plus/assets/11434289/89da4f6f-910d-41bb-a2a1-8ada9df785c9)

I mean the more cores the slower. For example: two computers both have 512G RAM, one has 32 cores and the other has 128 cores, when testing test 0, the 32 core one is more faster than the 128 core one. Because this test is in SEQ mode, I think each core has to traverse all address spaces in non-cacheable mode one by one, so it is slower.

**Comment by kilaterlee** [comment]
Do we need floating point numbers in the future? In LoongArch, the exception handler will save all registers such as GP, FP, and CSR. If we don't need floating point numbers in the future, I will consider removing the logic of saving FP.

**Comment by kilaterlee** [comment]
> Do we need floating point numbers in the future? In LoongArch, the exception handler will save all registers such as GP, FP, and CSR. If we don't need floating point numbers in the future, I will consider removing the logic of saving FP.

Ha, I see that calculating CPU clocks uses floating point numbers, so it seems like I still need to save FP in the exception handler.

**Comment by debrouxl** [comment]
Indeed, multiple locations in the code base use FP operations. That said, the x86/x86-64 exception handler does not save and restore FP registers, so I'd except that the code does not depend on them being preserved across exceptions - all the more most exceptions are fatal anyway.

**Comment by martinwhitaker** [comment]
Yes, for x86 and x86_64 the only exceptions that return are NMI and GPF after a RDMSR instruction. The exception handling code for those two cases is contained entirely within the `interrupt()` function and uses no floating point operations, so it shouldn't be necessary to preserve the floating point registers in the assembly language trap handler. However, taking a fresh look at the code, we probably should be preserving R8-R11 in case the `interrupt()` function happens to use those registers

**Comment by martinwhitaker** [comment]
I would just add R8-R11 to the registers pushed/popped in the `int_handler` code in startup64.S (and update the `interrupt()` function to be aware of the new stack layout).

**Comment by dchard** [comment]
I am not sure if this is the same issue or not, but with 7.20 on a DL360 Gen11 the first test is still super slow. 2S server with Xeon Gold 6430 and 256GB of DDR5 memory (Samsung M321R4GA3BB6-CQKVS). It was not able to pass the first test even after 5 minutes. The server is in default state, as it came out of HPE, nobody touched any BIOS settings.

With SMP Disbaled, it is fine.
With SMP Enabled both the init and test #0 takes forever:

![Memtest_7 20_test_0](https://github.com/user-attachments/assets/fa85dfda-5fca-4ef0-9d00-c31a8bc60f36)
![Memtest_7 20_PAR_Init](https://github.com/user-attachments/assets/95fa9e43-4ede-49a4-a63d-d6f7ffb65e6d)


Same machine with Memtest86 10.7 Pro runs these tests just fine.

**Comment by debrouxl** [comment]
Thanks for the report. That's a relatively nice computer :)

For testing purposes, is the init time significantly shorter if you 1) disable half of the cores and threads of your processors, and 2) re-enable all cores but disable hyperthreading (SMT) ? I don't have access to a DL360/DL380 Gen11 to check the exact procedure, but my DL380 Gen9 can do both through the same BIOS/Platform Configuration (RBSU) -> System Options -> Processor Options menu. ISTR that the Gen10 is similar.

**Comment by dchard** [comment]
@debrouxl it is indeed :-) Although I still have mixed feelings about these P-E core architecture on a server. We are going to do low latency networking, so HT and E cores are not much use to use.

The drive diag finishes in a couple hours, then I can try it. If you have any more idea you want me to test, this weekend is the best bet, maybe early next week, then it will go into prod.

**Comment by debrouxl** [comment]
Thanks for your tests.
Running tests 0, 1 and 2 on all cores ensures that more than core 0 (the BSP) can reach all memory... but we're aware that it's likely that computers which can't even do that have serious problems which are likely to be found through other means.

On both of your screenshots, memtest86+ indicates "SMP: Disabled", which isn't going to help testing speed. Was that just a result of disabling HT ?
Also, since you're using 7.20, could you enable HT again and pass the "numa" parameter to memtest86+ ? It won't change the init speed, but it should make it possible to lift the inter-socket bandwidth bottleneck (UPI link(s)), though that is less of an issue on 2S machines than on 4S machines whose processors have only two QPI/UPI links (e.g. SNB-EP to BDX-EP, or lower-grade Xeon SP), and allow testing at higher memory bandwidth.
Thanks in advance.

**Comment by dchard** [comment]
@debrouxl 

Happy to help.

> On both of your screenshots, memtest86+ indicates "SMP: Disabled", which isn't going to help testing speed. Was that just a result of disabling HT ?

Yes, it was only HT disabled in the BIOS, no other changes, Memtest86+ was started the regular way (SMP enabled). But wait a minute: it say 32 cores, 64 thread, SMP disabled and using: all cores?? First of all, disabling HT gives 64 total physical cores, not 32. Something is weird here. With HT disabled in the BIOS this should say: 64 cores, 64 threads, SMP disabled using: all cores.

I will try to set the NUMA parameter and get back to you.

**Comment by martinwhitaker** [comment]
Also try the `powersave=off` option.

**Comment by dchard** [comment]
@debrouxl 

Doing HT enabled and numa did not changed anything:

![Memtest_7 20_HT_enabled_NUMA](https://github.com/user-attachments/assets/524eb815-99ed-4b8e-9d12-2ac5d91f5c09)

@martinwhitaker 

Using powersave=off also did not changed anything:

The per core address tests are very slow in all cases. And when I disable HT, half the cores are gone the Memtest is not able to detect that this is a 2S server.

**Comment by debrouxl** [comment]
NUMA awareness won't change tests 0, 1 and 2, but it should speed up the parallel tests.

**Comment by dchard** [comment]
> NUMA awareness won't change tests 0, 1 and 2, but it should speed up the parallel tests.

Yeah, I understand that. Problem is it takes forever to pass tests 0,1,2 in order to get to the tests at hand :-)

It is still a question why memtest not indicating the "2S" part and showing only half the physical cores when HT is disabled.

**Comment by martinwhitaker** [comment]
> It is still a question why memtest not indicating the "2S" part and showing only half the physical cores when HT is disabled.

It will be a problem in parsing the ACPI tables. Could either be a bug/limitation in memtest86+ or a bug in the firmware that creates them.

As `powersave=off` made no difference, this is a different issue to the one reported by kilaterlee.

**Comment by debrouxl** [comment]
Right. `acpidump` from acpica-tools will provide a dump of the ACPI tables; among them, the MADT (APIC) and SRAT will be of main interest. I'm curious about the SLIT's contents as well, as I've never had the opportunity to get one from a SPR-SP 2S platform, but memtest86+ doesn't currently need to use it.

If possible, you could force disable the x2APIC mode in the UEFI settings. memtest86+ does not support it yet, and it's one thing which can disable SMP. Not sure why disabling HT would have to automatically force x2APIC mode on, though.

**Comment by x86fr** [comment]
Sapphire Rapids is supposed to work only in x2APIC by default, so many issues can come from the use of legally APIC, unfortunately.

**Comment by dchard** [comment]
> Right. `acpidump` from acpica-tools will provide a dump of the ACPI tables; among them, the MADT (APIC) and SRAT will be of main interest. I'm curious about the SLIT's contents as well, as I've never had the opportunity to get one from a SPR-SP 2S platform, but memtest86+ doesn't currently need to use it.

On it.

> If possible, you could force disable the x2APIC mode in the UEFI settings

This is not possible. I only have "auto" or "force enabled" in the BIOS.

**Comment by dchard** [comment]
@debrouxl 

Here is the full dump of the Dl360 Gen11:

[dl360-gen11-acpidump.zip](https://github.com/user-attachments/files/18188503/dl360-gen11-acpidump.zip)

Let me know if you need anything else, or you want me to try something.

**Comment by debrouxl** [comment]
Thanks :)
Several quick notes:
* in the SLIT, I see 10/20, which isn't unusual;
* in the SRAT, there's no Processor Local x2APIC Affinity subtable, so even if we supported x2APIC, the ACPI tables aren't telling us to use it...
* in the MADT and SRAT, all APIC / LAPIC entries are enabled, so HT is enabled, isn't it ?

**Comment by dchard** [comment]
>     * in the SLIT, I see 10/20, which isn't unusual;

Have no idea what this means. You want me to check a particular BIOS setting maybe?

>     * in the SRAT, there's no Processor Local x2APIC Affinity subtable, so even if we supported x2APIC, the ACPI tables aren't telling us to use it...

This is how the CPU is configured:

![image](https://github.com/user-attachments/assets/c5be6d23-95c6-4dec-a7bf-45bcff260e46)

This is what the BIOS say about the x2APIC options:

![image](https://github.com/user-attachments/assets/1cec97ea-016e-4667-bf7c-21c64384c34f)

Energy and power options:

![image](https://github.com/user-attachments/assets/aaec7dfc-02ff-4eec-8ea9-e54af38dc218)

>     * in the MADT and SRAT, all APIC / LAPIC entries are enabled, so HT is enabled, isn't it ?

Yes, HT was enabled this time.

**Comment by debrouxl** [comment]
About the SLIT: I just wondered which were the values representing relative local and remote NUMA node access latency advertised by the firmware. Remote access 20/10 = 2x slower than local access isn't unusual.

Would you be so kind as to provide ACPI dumps when 1) HT is disabled, and 2) additionally, Sub-NUMA Clustering, which appears in one of your screenshots, is enabled ? Besides, if you haven't done it yet, you may want to benchmark your desired workload with SNC enabled, as it can make some workloads slightly more efficient... or slightly less efficient.

**Comment by dchard** [comment]
> Would you be so kind as to provide ACPI dumps when 1) HT is disabled, and 2) additionally, Sub-NUMA Clustering, which appears in one of your screenshots, is enabled ?

Would you like the sub-NUMA part enabled with or without HT enabled?

> Besides, if you haven't done it yet, you may want to benchmark your desired workload with SNC enabled, as it can make some workloads slightly more efficient... or slightly less efficient.

The workload is low latency networking, the two CPUs each have a 25gig NIC which will be passed through to each VM and bind together in a NUMA aware fashion to make sure there is no traffic via QPI. Nor HT nor SpeedStep is not really an option for us as these are "slow" and involve a lot of context switching. I am not even sure how well the SW will handle the E cores...

**Comment by debrouxl** [comment]
> Would you like the sub-NUMA part enabled with or without HT enabled?
I had HT disabled (since I understand that it's your target configuration) + SNC enabled in mind, but HT enabled + SNC enabled should work too, since we'll already have the HT enabled and HT disabled configurations separately.

Ffor low-latency workloads, turbo overclocking might help, but downclocking would suck, indeed. And the better NUMA awareness provided by SNC may gain several percents.
And yeah, I'd say boo E cores for most workloads. They're much better performance/power tradeoffs than P cores, but still significantly lower performance than P cores. In order to obtain more consistent performance on a workload which is memory-bound anyway, I disabled both E cores and HT in an Alder Lake computer, which means that only 6 of the 20 HW threads are enabled... AMD's approach with Zen 4/4c and 5/5c cores, and the same instruction set between performance and efficiency cores, makes more sense.

**Comment by dchard** [comment]
@debrouxl 

Oki, here is an ACPI dump with HT disabled and NUMA-SNC2 enabled:

[dl360-gen11-HT-disbaled-NUMA-SNC2_enabled.zip](https://github.com/user-attachments/files/18196146/dl360-gen11-HT-disbaled-NUMA-SNC2_enabled.zip)

**Comment by debrouxl** [comment]
Thanks :)

Questions and remarks:
* in that HT disabled + SNC2 configuration, does memtest86+ run in SMP mode ? There's still no Processor Local x2APIC Affinity subtable in the SRAT, which I'd have expected to be shown when x2APIC mode is enabled (and memtest86+ then disables SMP).
* as before, the disabled cores are simply not shown in the SRAT and MADT, instead of being Enabled: 0. Both are valid, of course.
* the SLIT and SRAT show 4 NUMA nodes instead of 2, and 10/20/30/30 indicative distance. Since you're setting proper affinity, your workload may well benefit from SNC, yeah. I'd also expect the usage of 8 RAM sticks instead of 4 on each socket (reaching 1 DIMM per 2x32-bit channel) to help, at least it definitely would with the 1x64-bit channels of DDR4 or older.
3 different distance values is what I get on my 4S Opteron Interlagos & Abu Dhabi servers with 3 HT links per processor (8 NUMA nodes, distances 10/16/22) and 4S IVB-EP server with 2 QPI links per processor (4 NUMA nodes, 10/20/30).

**Comment by dchard** [comment]
>     * in that HT disabled + SNC2 configuration, does memtest86+ run in SMP mode ? There's still no Processor Local x2APIC Affinity subtable in the SRAT, which I'd have expected to be shown when x2APIC mode is enabled (and memtest86+ then disables SMP).

To be precise, this is in HT disabled + SNC2 configuration + x2APIC set to "auto" (default setting). With these settings Memtest86+ does SMP: 64 (PAR), still very slow Test0,1,2 (no "numa" flag passed), but it still indicates 32 cores and 64 threads, while it should be 64 cores and 64 threads and it is also not indicating 2S. Changing the x2APIC mode from "auto" to "force enabled" does change the Memtest86+ behavior: now it say SMP is disabled.

**Comment by dchard** [comment]
@debrouxl anything else I can do for you? More dumps you might need?

On the own address test slowness: the commercial counterpart does it only on core #0, and not on every core available. That is the reason it is faster. I have no idea if that is adequate enough or not, but non the less that seems to be the difference, and not numa or HT.

**Comment by debrouxl** [comment]
NUMA awareness has no effect on tests 0, 1 and 2, which are run on a single core at a time. The one thing which can influence the speed of tests 1 and 2 on some micro-architectures is the usage of non-temporal stores. It's not part of the main branch, though, because of the limited improvement, and untested impact on detection reliability.

Running tests 0, 1 and 2 only on core 0, i.e. the bootstrap processor certainly, does not ensure that at least one core on every socket, let alone every memory controller, can access all memory. Running these tests on all cores ensures that. Neither is optimal, since the former has insufficient coverage and the latter has excessive coverage. However, both methods are definitely simpler than trying to programmatically figure out the sets of cores corresponding to a given unit of the right granularity (I'd say memory controller, but given the variety of micro-architectures and MCs produced over time, I could be wrong) and test a single core from every set :)

> it still indicates 32 cores and 64 threads, while it should be 64 cores and 64 threads and it is also not indicating 2S

Point taken.

**Comment by x86fr** [comment]
> @debrouxl anything else I can do for you? More dumps you might need?
> 
> On the own address test slowness: the commercial counterpart does it only on core #0, and not on every core available. That is the reason it is faster. I have no idea if that is adequate enough or not, but non the less that seems to be the difference, and not numa or HT.

I 100% approve the change to core #0 only. IMO, that's the way to go because running it on all cores is useless.

**Comment by dchard** [comment]
> Running tests 0, 1 and 2 only on core 0, i.e. the bootstrap processor certainly, does not ensure that at least one core on every socket, let alone every memory controller, can access all memory. Running these tests on all cores ensures that. Neither is optimal, since the former has insufficient coverage and the latter has excessive coverage. However, both methods are definitely simpler than trying to programmatically figure out the sets of cores corresponding to a given unit of the right granularity (I'd say memory controller, but given the variety of micro-architectures and MCs produced over time, I could be wrong) and test a single core from every set :)

Exactly what came to my mind: test core #0 on each socket in case a 2S/4S system is detected. This might also be useful to fix the 2S/4S detection and core/thread numbers on N-way systems (with or without HT).

**Comment by debrouxl** [comment]
We do now have most of the pieces of infrastructure for running tests 0, (1) and 2 at 1 core per NUMA node... but yeah, 1 core per socket would be a reasonable tradeoff - and besides, it would handle the pre-NUMA multi-socket machines, too, at an additional complexity I'd expect to be low enough (maybe I'm wrong about that, though).

I like 4S workhorses because some non-distributable NUMA-aware workloads with large working sets I care about benefit from the additional memory channels and cores (at a given generation), but computers which have more than two sockets are a tiny minority in the real world. AMD stopped releasing new processors with > 2S scalability over a decade ago, and Intel does no longer provide > 2S capable processors at every generation.

**Comment by dchard** [comment]
@debrouxl "same" issue is present on a DL325 gen11 with a single AMD EPYC 9534and 512GB RAM: very slow own address test.

**Comment by x86fr** [comment]
Should be solved now
